### ISSUE TITLE
[Android] Suggestion: be able to exclude some assemblies from e4k stubs generation

### ISSUE THREAD

**Issue by postromantic** [mono/Embeddinator-4000#454]
e4k scans ALL referenced assemblies for public Activities etc. Usually it's not needed and cause problems. It would be nice to be able to exclude some assemblies from e4k java stubs generation via cli argument.

__Repro__
Create Xamarin Android library, add `Xamarin.GooglePlayServices.Location` nuget, create dummy class
```csharp
public class Class4 : Java.Lang.Object
{
    public void Test()
    {
        Console.WriteLine(typeof(GoogleApiClient));
    }
}
```
e4k will fail with bunch of errors for `com.google.android.gms.*` objects

**Comment by jonathanpeppers** [comment]
This is also the blocker to getting a Xamarin.Forms project working.

The issue is:
- We have to run `GenerateJavaStubs` on all user assemblies, because there could be an Android library project containing an `Activity` that needs to be added to the manifest and a Java stub generated so Java can create it. `GenerateJavaStubs` does `AndroidManifest.xml` generation, too.
- But we don't really need to generate Java for *every* `Java.Lang.Object` subclass. Like in the case of the support library/google play services or something like that. It seems like it could only generate java stubs for the assemblies passed to Embeddinator.

I think I need to talk to the Xamarin.Android team about this and see what they think.

**Comment by jonathanpeppers** [comment]
@postromantic can you try what I have on PR #460?

I changed so much stuff, I wanted to make sure it solves your problem above but doesn't break anything that used to work.

I was able to get this Xamarin.Forms app to work: https://github.com/davidortinau/build2017-new-in-xamarin-forms

**Comment by postromantic** [comment]
@jonathanpeppers `android-resource-designer` branch doesn't work for me. Even empty Xamarin Android library project fails now on LinkAssemblies task with
`Package.proj(21,5): error MSB4018:
Java.Interop.Tools.Diagnostics.XamarinAndroidException: error XA2006: Could not resolve reference to 'System.C
odeDom.Compiler.CompilerError' (defined in assembly 'CppSharp.Generator, Version=0.0.0.0, Culture=neutral, Pub
licKeyToken=null') with scope 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. Whe
n the scope is different from the defining assembly, it usually means that the type is forwarded. ---> Mono.Ce
cil.ResolutionException: Failed to resolve System.CodeDom.Compiler.CompilerError`

It start failing after this commit: https://github.com/mono/Embeddinator-4000/pull/460/commits/6609b86ed24bcd00872092622747b6f38302e0d1 (https://github.com/mono/Embeddinator-4000/pull/460/commits/059cf92ba9130488c15cac0765f0db368f95e637 works fine). OS: Windows

**Comment by jonathanpeppers** [comment]
Ok, is your target assembly in the same directory as Embeddinator?

If so, make a directory called `target` and move your C# assemblies in there.

So a structure like:
```
MonoEmbeddinator.exe
target\YourDLL.dll
target\OtherDependencyDLLs.dll
```
I usually just point to the `bin\Release` directory of the other project instead.

Then if you run:
```
MonoEmbeddinator.exe -gen=Java -platform=Android -c -o output target\YourDLL.dll
```
Maybe it works?

Since this caused you an issue, I probably need to think more about 6609b86 merging this PR. I did it in order to fix a Xamarin.Forms app I was testing.

**Comment by postromantic** [comment]
@jonathanpeppers, right=) i see you adding all dlls from main assembly dir.
it works this way

**Comment by postromantic** [comment]
however, original issue is still there. got bunch of java compiler errors like:
`output\src\mono\com\google\android\gms\tasks\OnFailureListenerImplementor.java:8: error: package com.google.android.gms.tasks does not exist
                com.google.android.gms.tasks.OnFailureListener
                                            ^
output\src\mono\com\google\android\gms\tasks\OnSuccessListenerImplementor.java:8: error: package com.google.android.gms.tasks does not exist
                com.google.android.gms.tasks.OnSuccessListener
`

**Comment by jonathanpeppers** [comment]
Ok, I will add a test for `Xamarin.GooglePlayServices.Location` and see if I see the same thing.

**Comment by postromantic** [comment]
Probably it's not the best nuget to test cause it has ~10 dependent nugets.  I think `Xamarin.Android.Support.v4` should also fail

**Comment by jonathanpeppers** [comment]
The one I got to work had a few: https://github.com/davidortinau/build2017-new-in-xamarin-forms/blob/master/Weather/WeatherApp/WeatherApp.Droid/packages.config

But Google Play Services is alot bigger than this.

**Comment by postromantic** [comment]
Yep, support packages works fine (in your example and on dummy projects), so it's play services only

**Comment by jonathanpeppers** [comment]
@postromantic I took out 6609b86 since the Xamarin.Forms sample still worked without it.

I will keep this issue open and address Google Play Services in a future PR.

Thanks for testing this out for us!

**Comment by postromantic** [comment]
@jonathanpeppers i think you should bring back https://github.com/mono/Embeddinator-4000/commit/6609b86ed24bcd00872092622747b6f38302e0d1 cause it solves another issue: be able to use resources from referenced assemblies (otherwise they are not in R.txt).
thanks

**Comment by jonathanpeppers** [comment]
When you were trying this, was the referenced assembly included in your AAR file:
```
zipinfo yourAARfile.aar
...
assets/assemblies/yourReferencedAssembly.dll
```

If it wasn't, it might be the linker removing the assembly entirely. The fix for that is to just make an empty static void method and call it from somewhere. The SQLite.raw library we were testing has something like that [here](https://github.com/mono/Embeddinator-4000/blob/master/tests/managed/android/android.cs#L284).

**Comment by jonathanpeppers** [comment]
@postromantic ok, I have a PR I think _might_ solve your issue.

Would you be able to try #463 and take a look at docs I wrote [here](https://github.com/jonathanpeppers/Embeddinator-4000/blob/android-dependencies/docs/getting-started-java-android.md#multiple-assemblies)?

If you still can't get your project to work, would you be able to send me your full project somehow?

**Comment by postromantic** [comment]
@jonathanpeppers, it works now, but there is one small issue =)

you've added a bunch of google play services to the test project except the one i need (Xamarin.GooglePlayServices.Location for geolocation capturing).  i'm using v29.0.0.1. e4k fails with:
`
AndroidManifest.xml(12): error APT0000:  Error: No resource found that matches the given name (at 'value' with value '@integer/google_play_services_version').
`

So it works only after i manually created integer resource in my assembly like
```xml
<integer name="google_play_services_version">8487000</integer>
```
(magic number is from another app with same play services)

**Comment by jonathanpeppers** [comment]
Glad you got it to work!

Yeah your issue makes sense, because it's not going to include the resources from Google Play Services.

In the future, we are looking to add a `-ref` command line switch to pass Embeddinator assemblies you want included (Android resources, etc.), but you don't want Embeddinator to generate Java output for. I think this will address your new issue, but it might be a while before we get that feature.

When I get back to editing the docs, I'll put an example of what you had to do here.

I'll close this for now, because I think it solves your original problem. Let us know if you find something else!

**Comment by postromantic** [comment]
Thanks @jonathanpeppers!
